httpcore/_async/http_proxy.py

```diff
@@ -140,47 +140,52 @@ def __init__(
         connection = await self._get_connection_from_pool(origin)
 
         if connection is None:
-            connection = AsyncHTTPConnection(
-                origin=origin, http2=False, ssl_context=self._ssl_context,
+            # First, create a connection to the proxy server
+            proxy_connection = AsyncHTTPConnection(
+                origin=self.proxy_origin, http2=False, ssl_context=self._ssl_context,
             )
             async with self._thread_lock:
                 self._connections.setdefault(origin, set())
-                self._connections[origin].add(connection)
+                self._connections[origin].add(proxy_connection)
 
-            # Establish the connection by issuing a CONNECT request...
+            # Issue a CONNECT request...
 
             # CONNECT www.example.org:80 HTTP/1.1
             # [proxy-headers]
             target = b"%b:%d" % (url[1], url[2])
             connect_url = self.proxy_origin + (target,)
-            connect_headers = self.proxy_headers
-            proxy_response = await connection.request(
+            connect_headers = self.proxy_headers or [(b"host", self.proxy_origin[1])]
```

Comment:
The `host` header is required by h11 but not typically passed as part of the proxy headers.

Not sure if we want to merge the headers here or leave that to the caller to sort out.

Reply:
A very obvious `h11._util.LocalProtocolError: Missing mandatory Host: header`.

Looking at [the old `proxy_http` module](https://github.com/encode/httpx/blob/631ba97635262fc72556a1638bac42cc843fe2ea/httpx/_dispatch/proxy_http.py) I can't see we were adding it ourselves, although we do take care of the proxy authentication, which I guess we can take care of later.

I actually don't see us adding the `host` header anywhere else...

Reply:
Yes, I went across that one while debugging.

I guess since it's a wrinkle with `h11`, and not something at the HTTP spec level, then `httpcore` should handle it and add the `host` header by default (but _only_ for the `CONNECT` call). Thoughts?

Reply:
Looks like in standard requests it's added by the [`Request` model in HTTPX](https://github.com/encode/httpx/blob/master/httpx/_models.py#L645) and passed in as expected, however when using a [Proxy](https://github.com/encode/httpx/blob/master/httpx/_config.py#L326-L354) it is not set.

Maybe for consistency we should probably create those in Proxy and pass them to `httpcore`?

Reply:
Yes, we can probably have `Proxy.headers` and pass them when calling `httpcore`, that sounds nice.

Reply:
I noticed that when forwarding the proxy and request headers were merged, so the `host` header can simply be taken from the request headers.